### ISSUE TITLE
Dask DataFrame set_index does not work when column name matches index name

### ISSUE THREAD

**Issue by georgek-auro** [dask/dask#9331]
**What happened:**

Calling set_index providing a column name that matches the name of the index does not perform the set_index.

**What you expected to happen:**

The provided column becomes the index, as was the case prior to Dask==2022.05.1

**Minimal Complete Verifiable Example:**

```python
import dask.dataframe as dd
import pandas as pd

df = dd.from_pandas(pd.DataFrame([
    {'a': 1, 'b': 2},
    {'a': 2, 'b': 1}]
), npartitions=1)
df = df.set_index('a')
df = df.rename(columns={'b': 'a'})
df.set_index('a').head()
```

**Anything else we need to know?:**

This is the [branch](https://github.com/dask/dask/blob/0b637ab775fa177d0af737e4648a89484aee5c74/dask/dataframe/core.py#L4816) responsible for the behavior, it was introduced in a commit that seems [unrelated](https://github.com/dask/dask/commit/d458533b5f04494c8eb070adaff6470cea2cf6b0). Dask versions prior to 2022.05.1 are unaffected.

**Environment**

Dask version: 2022.5.1
Python version: 3.10.5
Operating System: Ubuntu 20.04
Install method (conda, pip, source): pip

**Comment by quasiben** [comment]
At the moment it looks like Dask chooses to do nothing if the column name matches the index name:
https://github.com/dask/dask/blob/0b637ab775fa177d0af737e4648a89484aee5c74/dask/dataframe/core.py#L4813-L4817

@jorloplaz do you have any thoughts whether this optimization was intended for a particular workflow ?

**Comment by jorloplaz** [comment]
IMHO, a DataFrame should never have two columns with the same name, and this should be controlled at all times (see the discussion regarding #9122). This includes as a particular case the one currently acting as the index, so in your example the call to  `rename` should raise an error, as there would be two columns called `'a'`.  

Think about it in simpler terms: if you want to use `b` as index, why don't you use `set_index('b')` directly?

Assuming the above holds, once you have a column acting as index, calling `set_index` on that same column should do nothing (in string form or `Series` form; so doing `set_index(df.index)` should also be a no-op). 

The only particular case where it could trigger some action is if the index is found to be unsorted on runtime (e.g., because `known_divisions=False` or `df.index.is_monotonic_increasing` happens to be `False`). 

@jsignell @pavithraes Any thoughts on this?

**Comment by jorloplaz** [comment]
I don't think that solves the problem. What happens above if you do first `reset_index()` and then `set_index('a')`. My guess is that, since there's another column called `'a'` when resetting, it is reset with the name `index`, and then you end up with two columns `'a'` and `'index'` instead of `'b'` and `'a'`, which was the original situation.

Even if Dask doesn't currently prevent you now from having multiple columns with the same name, there's no good reason why you should do it. In my view, name clashes show a flaw in the processing pipeline.

Pandas does allow non-unicity because in any moment you can transpose the DataFrame with `.T`, so columns can become the index, and the index can for sure have duplicate values. But Dask is very different, because it is on the index that is partitioned, but not in the columns, so a frame isn't transposable (i.e., rows and columns are not interchangeable).

**Comment by georgek-auro** [comment]
Why don't we give the user a warning when such a situation occurs as well as make a note in the documentation?

**Comment by jorloplaz** [comment]
Mmm, as I said, if you have a column `'a'` acting as index, it's because that's one of the original columns of your dataframe. So once it is the index, doing `.set_index(`'a'`)` on that frame makes sense it is a no-op (recall that `.set_index()` requires computing quantiles and shuffling, which is expensive, so it should be done with care).

If you do have another column in the frame with that same name, it shows some error on your side. Because that wouldn't be possible from the beginning: if you load your frame from an SQL table, there cannot be any columns with the same name, same goes for Parquet files, etc.

So you did something between loading your data and calling `.set_index()` that introduced a duplicate column. In the snippet above it is `.rename()`. Why would you do such a thing? I cannot think of any good reason...

**Comment by jsignell** [comment]
I agree in principle with @jorloplaz that this is an odd behavior and that duplicate column names are never a good idea. However since they are still possible to create, it makes sense to at least give a warning in this case.

**Comment by jorloplaz** [comment]
Thank you @jsignell ! As for me, I'll start working on preventing duplicate column names.

**Comment by jorloplaz** [comment]
I opened a PR that prevents you from creating frames with duplicate column names.
